beacon-chain/forkchoice/protoarray/proposer_boost_test.go

```diff
@@ -0,0 +1,468 @@
+		//         0
+		//         |
+		//         1 <- HEAD
+		slot := types.Slot(1)
+		newRoot := indexToHash(1)
+		require.NoError(t,
+			f.ProcessBlock(
+				ctx,
+				slot,
+				newRoot,
+				headRoot,
+				graffiti,
+				jEpoch,
+				fEpoch,
+			),
+		)
+		f.ProcessAttestation(ctx, []uint64{0}, newRoot, fEpoch)
+		headRoot, err = f.Head(ctx, jEpoch, zeroHash, balances, fEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, newRoot, headRoot, "Incorrect head for justified epoch at slot 1")
+
+		// Insert block at slot 2 into the tree and verify head is at that block:
+		//         0
+		//         |
+		//         1
+		//         |
+		//         2 <- HEAD
+		slot = types.Slot(2)
+		newRoot = indexToHash(2)
+		require.NoError(t,
+			f.ProcessBlock(
+				ctx,
+				slot,
+				newRoot,
+				headRoot,
+				graffiti,
+				jEpoch,
+				fEpoch,
+			),
+		)
+		f.ProcessAttestation(ctx, []uint64{1}, newRoot, fEpoch)
+		headRoot, err = f.Head(ctx, jEpoch, zeroHash, balances, fEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, newRoot, headRoot, "Incorrect head for justified epoch at slot 2")
+
+		// Insert block at slot 3 into the tree and verify head is at that block:
+		//         0
+		//         |
+		//         1
+		//         |
+		//         2
+		//         |
+		//         3 <- HEAD
+		slot = types.Slot(2)
+		newRoot = indexToHash(2)
+		require.NoError(t,
+			f.ProcessBlock(
+				ctx,
+				slot,
+				newRoot,
+				headRoot,
+				graffiti,
+				jEpoch,
+				fEpoch,
+			),
+		)
+		f.ProcessAttestation(ctx, []uint64{2}, newRoot, fEpoch)
+		headRoot, err = f.Head(ctx, jEpoch, zeroHash, balances, fEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, newRoot, headRoot, "Incorrect head for justified epoch at slot 3")
+
+		// Insert a second block at slot 3 into the tree and boost its score.
+		//         0
+		//         |
+		//         1
+		//         |
+		//         2
+		//        / \
+		//       3   4 <- HEAD
+		slot = types.Slot(3)
+		newRoot = indexToHash(4)
+		require.NoError(t,
+			f.ProcessBlock(
+				ctx,
+				slot,
+				newRoot,
+				headRoot,
+				graffiti,
+				jEpoch,
+				fEpoch,
+			),
+		)
+		f.ProcessAttestation(ctx, []uint64{3}, newRoot, fEpoch)
+		threeSlots := 3 * params.BeaconConfig().SecondsPerSlot
+		genesisTime := time.Now().Add(-time.Second * time.Duration(threeSlots))
+		require.NoError(t, f.BoostProposerRoot(ctx, slot, newRoot, genesisTime))
+		headRoot, err = f.Head(ctx, jEpoch, zeroHash, balances, fEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, newRoot, headRoot, "Incorrect head for justified epoch at slot 3")
+
+		// Check the ancestor scores from the store.
+		require.Equal(t, 4, len(f.store.nodes))
+		require.Equal(t, f.store.nodes[0].weight, uint64(0)) // Genesis has 0 weight here.
+
+		// Expect nodes to have a boosted, back-propagated score.
+		require.Equal(t, f.store.nodes[1].weight, uint64(47))
+		require.Equal(t, f.store.nodes[2].weight, uint64(37))
+		require.Equal(t, f.store.nodes[3].weight, uint64(17))
```

Comment:
I don't follow this. Genesis is the oldest block at index `0` and has `0` score - makes perfect sense. But then the next oldest block at index `1` has score `47`, and the following blocks' scores are getting lower? Something is backwards here...

Reply:
Ancestors have the added weights of their children. Genesis is a special exception at 0 weight, but otherwise assuming a block, A, that is not-genesis:

A -> B -> C

Where each one has a weight of 10 individually, the final weights will look like
```
(A: 30) -> (B: 20) -> (C: 10)
```

The boost adds 7 to the weight, so if C is boosted, we would have 
```
(A: 37) -> (B: 27) -> (C: 17)
```

In this case, we have a small fork:

```
(A: 47) -> (B: 37) -> (C: 17)
                  \_->(D: 10)
```
So B has its own weight, 10, and the sum of of both C and D thats why you see that weight `37` in the middle instead of the normal progression of (37 -> 27 -> 17)

Reply:
This feels like good commentary to add in code